### PR TITLE
Updated wipe logic to respect retraction speed during wiping

### PR DESCRIPTION
**Pull request by igiannakas**
This has been a fun journey! This PR replaces #3060 

**Recap of the issue:**

Observed that while performing filament changes the wipe logic on occasion results in the full filament swap retract value being used over a very small distance. This causes the extruder to stall resulting in filament blobs on the surface. This PR aims to improve the wipe logic making it more "extruder safe".

image

This issue has been present for a while; however was not manifesting itself in the prints as the previous version of the ensure vertical thickness algorithm & detect narrow internal infill algorithms were not creating small fills that were extruded last in the layer.

This issue also occurs without filament changes taking place; however as the retraction distance for regular printing is far smaller, the issue doesn't appear as the extruder should not be getting over worked. However this may materialise in Bowden printers like this issue here: https://github.com/SoftFever/OrcaSlicer/issues/2993 and the last comment made to it

It may also cause issues with material like TPU and PETG and silk PLA that do not like being retracted quickly as they can be pulled apart from the nozzle and suck air in causing holes on outer wall surfaces when continuing to print.

This PR may also fix this issue here: https://github.com/SoftFever/OrcaSlicer/issues/2753 as it appears to be due to similar wipe over a small distance reasons. 

**Root Cause:**

The current wipe code does not respect the retraction speed set in the filament profile/extruder profile. Instead it wipes the requested retraction amount over the available wipe distance irrespective of how small it is, or how fast the tool head is moving. This means that for a tool change where significant retraction is requested, or if the user sets the wipe distance to be too small and the retraction distance to be too high, the extruder would attempt to retract the filament with significantly higher speeds than the set retraction speed.

**This PR aims to address this issue by:**
1. Calculating the maximum retraction length possible in the available wipe distance in order to obey the set by the user retraction speed.
```
                double maxRetractionLength = gcodegen.config().retraction_speed.get_at(gcodegen.writer().extruder()->id())
                                                                     * (wipe_path.length() / 1000000)
                                                                     / _wipe_speed;
```
The formula used above is derived by solving this equation:
- **Time needed to retract**= requested retraction during wipe / retraction speed
- **Time available to wipe** = available wipe length / wipe speed
- **Max Retraction Length** = requested retraction during wipe * (Time available to wipe / Time needed to retract)

With the simplified form of this equation being:

image

2. If the maximum retraction length possible is less than the user requested one, it performs an immediate retraction of the difference and then wipes with the remaining.
```
                    gcode += gcodegen.writer().retract(length - maxRetractionLength + dE_retracted, toolchange);
                    length = maxRetractionLength;
```
4. In essence it overrides the retract before wipe setting from the user in the scenarios where the requested retraction while wiping would breach the user selected retraction speed.

**Screenshots:**

**Wiping before filament change** - most of the filament amount is retracted prior to the wipe move as the wipe area is small
image

**Wiping during layer change** - depending on the user set wipe distance, retraction acceleration and wiping speed, it can fully wipe while retracting
image

**Retract before wipe is respected:**
The below is with 50% retraction (but it will do a double retract move if needed)
image
image

I think this change now makes wiping bullet proof :D 

@SoftFever  let me know your thoughts. I'll close down the other PR as this approach is much better and cleaner.

**Edit:** this method will also protect the user from a too short wipe distance set, or too small retraction before wipe %, compared to the filament retraction rate. It will perform a retraction before a wipe move even if the user has told it not to but the filament retraction speed cannot allow for it.

**Edit 13 Dec:** Made the above feature set an option. Removed the retract operation before wipe and placed after the wipe.

### CODE REVIEW

**Comment by igiannakas** [comment]
I'll remove the Gcode debug comments once review is concluded and its ready to publish.

**Comment by igiannakas** [comment]
I am still debating whether to retract the excess amount before the wipe or after. Keen to hear your thoughts on this @SoftFever as I am unsure what's the best approach.

Im noticing Prusa doing the retraction after the wipe if there is no area to complete the wipe on.
image

Prusa doesn't have a user configurable wipe distance, so it calculates the required wipe distance dynamically in order to respect the retraction speed - like what this PR does. If the available wipe distance is less than the required one, it wipes first with the available distance and allowed retraction amount, then retracts the rest.

I dont particularly like this approach of having a variable wipe distance, so it's a plus to have a user configurable setting for this in Orca. However I am unsure on whether to retract before or after the wipe move :S What would the benefit / con of each approach be?

I am thinking that retracting after the wipe move may be better as the nozzle doesn't stop to retract and then start again to wipe which may result in blobs if the extruder is not fast enough?

Thoughts?

**Comment by killerline10168** [comment]
I am not sure if this will even contribute one bit to this topic, but take a look at my test results in my ticket regarding wipes and their effect on walls.

https://github.com/SoftFever/OrcaSlicer/issues/1649

**Comment by igiannakas** [comment]
> I am not sure if this will even contribute one bit to this topic, but take a look at my test results in my ticket regarding wipes and their effect on walls.
> 
> #1649

Well partly - basically your tests indicate I think that we dont want to be retracting before the wipe?

**Comment by killerline10168** [comment]
> > I am not sure if this will even contribute one bit to this topic, but take a look at my test results in my ticket regarding wipes and their effect on walls.
> > #1649
> 
> Well partly - basically your tests indicate I think that we dont want to be retracting before the wipe?

With my experience regarding wipes, I've also come to the conclusion that wiping before retracting is the superior alternative. When the printer retracts before it wipes, it kinda dry-wipes the print, which gives a worse-looking seam and can even create gaps on the wiped surface. IMO the point of a wipe is to wipe the remaining plastic that is still pressurized in the nozzle, which also gives the added benefit of  "closing up" contours. When said plastic has depressurized and somewhat left the nozzle, then it is much easier to retract the filament successfully without stringing occurring. 

TL:DR: No, I do not think we should retract before we wipe. Although haven't scientifically tested this yet, just my 2 cents.

**Comment by igiannakas** [comment]
> > > I am not sure if this will even contribute one bit to this topic, but take a look at my test results in my ticket regarding wipes and their effect on walls.
> > > #1649
> > 
> > 
> > Well partly - basically your tests indicate I think that we dont want to be retracting before the wipe?
> 
> With my experience regarding wipes, I've come to the conclusion that wiping before retracting is the superior alternative. When the printer retracts before it wipes, it kinda dry-wipes the print, which gives a worse-looking seam and can even create gaps on the wiped surface. IMO the point of a wipe is to wipe the remaining plastic that is still pressurized in the nozzle, which also gives the added benefit of "closing up" contours. When said plastic has depressurized and somewhat left the nozzle, then it is much easier to retract the filament successfully without stringing occurring.
> 
> TL:DR: No, I do not think we should retract before we wipe. Although, I haven't scientifically tested this yet, just my 2 cents.

Cool. I’ve got a local build with the retraction happening after the wipe move. Seems to work well. If there is interest I can get a second PR with it and then we choose what is merged and whether we want this as an option in the gui.

**Comment by killerline10168** [comment]
> c

I am all for options and letting users decide, but maybe not on a SuperSlicer level. If wiping before retracting is the objectively better alternative, should we not just stick to that for now?  

Also, as you are already working with wipes, is it possible for you to take a look at my ticket? I believe this would further improve print quality from Orca. 

Thank you for all your hard work! Great community here!

**Comment by igiannakas** [comment]
Updated PR to make this whole feature an option under the extruder settings just in case a user wants to revert to the previous behaviour.

**Comment by SoftFever** [comment]
> This has been a fun journey! This PR replaces #3060
> 
> **Recap of the issue:**
> 
> Observed that while performing filament changes the wipe logic on occasion results in the full filament swap retract value being used over a very small distance. This causes the extruder to stall resulting in filament blobs on the surface. This PR aims to improve the wipe logic making it more "extruder safe".
> 
> image 
> 
> This issue has been present for a while; however was not manifesting itself in the prints as the previous version of the ensure vertical thickness algorithm & detect narrow internal infill algorithms were not creating small fills that were extruded last in the layer.
> 
> This issue also occurs without filament changes taking place; however as the retraction distance for regular printing is far smaller, the issue doesn't appear as the extruder should not be getting over worked. However this may materialise in Bowden printers like this issue here: #2993 and the last comment made to it
> 
> It may also cause issues with material like TPU and PETG and silk PLA that do not like being retracted quickly as they can be pulled apart from the nozzle and suck air in causing holes on outer wall surfaces when continuing to print.
> 
> This PR may also fix this issue here: #2753 as it appears to be due to similar wipe over a small distance reasons.
> 
> **Root Cause:**
> 
> The current wipe code does not respect the retraction speed set in the filament profile/extruder profile. Instead it wipes the requested retraction amount over the available wipe distance irrespective of how small it is, or how fast the tool head is moving. This means that for a tool change where significant retraction is requested, or if the user sets the wipe distance to be too small and the retraction distance to be too high, the extruder would attempt to retract the filament with significantly higher speeds than the set retraction speed.
> 
> **This PR aims to address this issue by:**
> 
> 1. Calculating the maximum retraction length possible in the available wipe distance in order to obey the set by the user retraction speed.
> 
> ```
>                 double maxRetractionLength = gcodegen.config().retraction_speed.get_at(gcodegen.writer().extruder()->id())
>                                                                      * (wipe_path.length() / 1000000)
>                                                                      / _wipe_speed;
> ```
> 
> The formula used above is derived by solving this equation:
> 
> * **Time needed to retract**= requested retraction during wipe / retraction speed
> * **Time available to wipe** = available wipe length / wipe speed
> * **Max Retraction Length** = requested retraction during wipe * (Time available to wipe / Time needed to retract)
> 
> With the simplified form of this equation being:
> 
> image
> 
> 2. If the maximum retraction length possible is less than the user requested one, it performs an immediate retraction of the difference and then wipes with the remaining.
> 
> ```
>                     gcode += gcodegen.writer().retract(length - maxRetractionLength + dE_retracted, toolchange);
>                     length = maxRetractionLength;
> ```
> 
> 4. In essence it overrides the retract before wipe setting from the user in the scenarios where the requested retraction while wiping would breach the user selected retraction speed.
> 
> **Screenshots:**
> 
> **Wiping before filament change** - most of the filament amount is retracted prior to the wipe move as the wipe area is small image
> 
> **Wiping during layer change** - depending on the user set wipe distance, retraction acceleration and wiping speed, it can fully wipe while retracting image
> 
> **Retract before wipe is respected:** The below is with 50% retraction (but it will do a double retract move if needed) image image
> 
> I think this change now makes wiping bullet proof :D
> 
> @SoftFever let me know your thoughts. I'll close down the other PR as this approach is much better and cleaner.
> 
> **Edit:** this method will also protect the user from a too short wipe distance set, or too small retraction before wipe %, compared to the filament retraction rate. It will perform a retraction before a wipe move even if the user has told it not to but the filament retraction speed cannot allow for it.
> 
> **Edit 13 Dec:** Made the above feature set an option. Removed the retract operation before wipe and placed after the wipe.

Thanks for the comprehensive summary and write up, which is exactly what was in my mind 👍

**Comment by igiannakas** [comment]
Your welcome! It's been fun trying to figure it out! 

There are a couple of things to agree if that's ok?
1.  whether to retract the excess before or after the wipe operation is performed. I've got the code developed for both
2. whether to make this "respect retraction speed when wiping" an option to the user or not? 

Personally I believe that we should retract after the wipe move as I would think a stationary nozzle when retracting defeats the purpose of the wipe move (and if the user wants the retract before wipe, they can set the corresponding parameter anyway).

And making it an option Im not sure - I can't see the reason when a user would not want this enabled but maybe I\m wrong...

**Comment by SoftFever** [comment]
> Your welcome! It's been fun trying to figure it out!
> 
> There are a couple of things to agree if that's ok?
> 
> 1. whether to retract the excess before or after the wipe operation is performed. I've got the code developed for both
> 2. whether to make this "respect retraction speed when wiping" an option to the user or not?
> 
> Personally I believe that we should retract after the wipe move as I would think a stationary nozzle when retracting defeats the purpose of the wipe move (and if the user wants the retract before wipe, they can set the corresponding parameter anyway).
> 
> And making it an option Im not sure - I can't see the reason when a user would not want this enabled but maybe I\m wrong...

Sure.
For the first question, it's better to retract the excess before the wipe move. retract after will defeat the purpose of "wipe".
For the second question, in my opinion, the option is not necessary since it's a bug-fixing/improvement. Therefore, the fail-safe strategy should always be enabled."

**Comment by igiannakas** [comment]
Gotcha. However if the user has not set a retract before wipe, we are ignoring their setting with this. So maybe if the user has enabled the retract before wipe, do it before, if they haven't, do it after?

I'll amend it like the above and remove the option from the extruder.

**Comment by SoftFever** [comment]
> Gotcha. However if the user has not set a retract before wipe, we are ignoring their setting with this. So maybe if the user has enabled the retract before wipe, do it before, if they haven't, do it after?
> 
> I'll amend it like the above and remove the option from the extruder.

 `has not set a retract before wipe` are you referring to this option when it's 0? if so, then 0 is not difference than other numbers here. We are handling it consistently(and logically).
image

I will do the code review tomorrow. It's a bit later here.
Have a good day/night ^^

**Comment by igiannakas** [comment]
I've pushed a new commit for the above. Take a look and let me know. Basically if the retract amount before wipe is >0 then do the excess retraction before the wipe as the user has indicated that they prefer to do a retraction before wipe. If this parameter is 0 the excess is retracted at the end of the wipe move.

File: src/libslic3r/GCodeWriter.cpp [hunk 1]
```diff
@@ -682,6 +682,15 @@ std::string GCodeWriter::extrude_to_xyz(const Vec3d &point, double dE, const std
     return w.string();
 }
 
+std::string GCodeWriter::retract(double retraction_length, bool toolchange)
```
Comment (SoftFever) [inline src/libslic3r/GCodeWriter.cpp:685]: Can you explain why you add this function?

File: src/libslic3r/GCodeWriter.cpp [hunk 2]
```diff
@@ -682,6 +682,15 @@ std::string GCodeWriter::extrude_to_xyz(const Vec3d &point, double dE, const std
...
+    return this->_retract(
+        retraction_length,
+        toolchange ? m_extruder->retract_restart_extra(): m_extruder->retract_restart_extra_toolchange(),
+        "retract"
+    );
+}
+
 std::string GCodeWriter::retract(bool before_wipe)
 {
     double factor = before_wipe ? m_extruder->retract_before_wipe() : 1.;
```
[the remark this answers is not in the document]
Comment (igiannakas) [reply src/libslic3r/GCodeWriter.cpp:696]: The problem there is that I don’t have the actual wipe path length or velocity available to do the calculation. Let me have a think about it, it may be a good idea!!

File: src/libslic3r/GCodeWriter.cpp (hunk 1 above)
Comment (igiannakas) [reply src/libslic3r/GCodeWriter.cpp:685]: Because there is no function to retract an arbitrary amount of filament :(

File: src/libslic3r/GCodeWriter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SoftFever) [reply src/libslic3r/GCodeWriter.cpp:696]: take a look at here, this is the only place when we call it with the parameter set to true, we can pass the wipe length here
And same for the retract_for_toolchange function.

```
std::string GCode::retract(bool toolchange, bool is_last_retraction, LiftType lift_type)
{
    std::string gcode;

    if (m_writer.extruder() == nullptr)
        return gcode;

    // wipe (if it's enabled for this extruder and we have a stored wipe path and no-zero wipe distance)
    if (EXTRUDER_CONFIG(wipe) && m_wipe.has_path() && scale_(EXTRUDER_CONFIG(wipe_distance)) > SCALED_EPSILON) {
        gcode += toolchange ? m_writer.retract_for_toolchange(true) : m_writer.retract(true);
        gcode += m_wipe.wipe(*this, toolchange, is_last_retraction);
    }
```

**Comment by igiannakas** [comment]
Refactored the code.

I've pulled out the length calculation to a dedicated function and then use it to trigger the right retractions and pass on the information to the wipe function. Looks a lot cleaner I think :) and hopefully easier to merge upstream changes in the future

File: src/libslic3r/GCodeWriter.cpp (hunk 1 above)
Comment (igiannakas) [reply src/libslic3r/GCodeWriter.cpp]: Removed it in latest build. Added optional variable to the existing functions.

File: src/libslic3r/GCodeWriter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (igiannakas) [reply src/libslic3r/GCodeWriter.cpp:687]: Done - moved the length calculation to a separate function so i can re use it in both the wipe logic and the retract before wipe call.

File: src/libslic3r/GCode.cpp [hunk 3]
```diff
@@ -295,7 +295,65 @@ static std::vector<Vec2d> get_path_of_change_filament(const Print& print)
...
+        wipe_speed = std::max(wipe_speed, 10.0);
+
+        // Process wipe path & calculate wipe path length
+        double wipe_dist = scale_(config.wipe_distance.get_at(extruder_id));
+        Polyline wipe_path = {last_pos};
+        wipe_path.append(this->path.points.begin() + 1, this->path.points.end());
+        double wipe_path_length = std::min(wipe_path.length(), wipe_dist);
+
+        // Calculate the maximum retraction amount during wipe
+        retractionDuringWipe = config.retraction_speed.get_at(extruder_id) * (wipe_path_length / 1000000) / wipe_speed;
```
Comment (SoftFever) [inline src/libslic3r/GCode.cpp:338]: Question: what does `wipe_path_length / 1000000)` do?
If we intended to unscale it, consider to use 'unscale_' macro instead.

Comment (igiannakas) [reply src/libslic3r/GCode.cpp:338]: Ah yes makes sense!! Thank you 👌🏻

File: src/libslic3r/GCodeWriter.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (SoftFever) [reply src/libslic3r/GCodeWriter.cpp:687]: > Done - moved the length calculation to a separate function so i can re use it in both the wipe logic and the retract before wipe call.

Nice.
The new changes are very clean now.

**Review by SoftFever — APPROVED** [review]
Looks good 👍

**Comment by SoftFever** [comment]
@igiannakas 
Thank you very much, merged.
Cheers!
